src/ModernGL.jl

```diff
@@ -59,35 +59,83 @@ isavailable(ptr::Ptr{Cvoid}) = !(
     ptr == convert(Ptr{Cvoid},  3)
 )
 
-abstract type Enum end
-
-macro GenEnums(list)
-    tmp = list.args
-    enumName = tmp[2]
-    splice!(tmp, 1:2)
-    enumType    = typeof(eval(tmp[4].args[1].args[2]))
-    enumdict1   = Dict{enumType, Symbol}()
-    for elem in tmp
-        if Meta.isexpr(elem, :const)
-            enumdict1[eval(elem.args[1].args[2])] = elem.args[1].args[1]
-        end
+struct GLENUM{Sym, T}
+    number::T
+    name::Symbol
```

Comment:
I'm not sure if I like this to be so heavyweight, with a symbol field and the Sym type parameter...
We already have massive compilation latency problems and putting the name into the type parameter means, that all code using GLENUM will try to specialize on `Sym`.
Not sure how big the problem is considering that there isn't much to specialize on.. But it still seems excessive!
Maybe it would be easier to group the enums, so that in one group they have their own type, and therefore don't overlap? I don't really see any end user package using the overload infrastructure - this is rather something that should be set in stone by ModernGL itself, or how do you plan to use this? 
But it does sounds confusing to me, that when you load some package your enums suddenly get renamed...
Also, does the symbol field actually work in all cases with the C interop? I'm not sure if there are cases where we need to pass an array of enums, or what other problems could arise.
Maybe we can use the implementation from https://github.com/JuliaInterop/CEnum.jl instead, and then put a bit of work into finding some groups, that we need to separate to not get name clashes?

Reply:
I was imagining that users might want to overload specific values to return whatever name is most relevant to the debugging work they're doing, but you're probably right that it's not a likely scenario in practice.

I'm not sure how much this increases the compilation times, as doesn't the compilation only happen after you first dispatch to it? I thought that in practice, this function is merely something you call once or twice in the REPL while debugging, or _maybe_ 3-10 times when intercepting OpenGL errors. Did I miss a use-case?

 I would guess that the project's huge compilation time comes from pre-emptively filling in the GLENUM lookup dictionary, because the GLENUM struct also has a type parameter for the Symbol, which means it's compiling hundreds of specializations upon loading the package. Speaking of which, is there a reason behind that, or would you be interested in a commit which removes it? It seems redundant anyway because it already keeps the Symbol as a field.

I'd be surprised if there are any cirumstances in which you'd send `GLENUM` to C. Wouldn't external libraries want plain `GLenum` values, rather than this package's special wrapper?

Reply:
@SimonDanisch Sorry, I think I misunderstood your original comment. I assumed you were talking about my new implementation of the `GLENUM` constructor, dispatching on `Val{Int(i)}`. But it sounds like you were talking about the first type parameter of `GLENUM{Sym, T}`? That was already there before I made these changes, and as I said above I agree we could get rid of it.

I'm still a bit confused on the circumstances where you'd want to pass a `GLENUM` into C, as opposed to a plain `GLenum`.

Also, I just checked over the code, and I was mistaken about compilation times. The lookup dictionary for enums doesn't contain a bunch of `GLENUM` instances, merely the symbols, so there's no excessive compilation there. The main cost of compilation is most likely the sheer overhead of setting up so many functions and constants, which is probably hard to avoid.

On the note of my new `GLENUM()` constructor, I just discovered that Julia has a `@nospecialize` macro, which should guarantee that the compiler only makes 3 overloads -- one for `Val{1}`, another for `Val{0}`, and then the catchall for any other `Val`.

**TL;DR** I'll cut out the `Sym` type parameter from `GLENUM`, and also throw `@nospecialize` in there.

Reply:
After updating the code, it takes about 5 seconds to precompile ModernGL on my machine, and then starting up a window (which involves a number of GL calls but also my own code) altogether takes ~3 seconds.

Reply:
Oh and as for grouping the GL enums so you can see all the names for each value, that would be nice, but I think it would need extra engineering work. A naive implementation would lead to creating *thousands* of tiny dictionaries, just for a little debugging helper! I think you'd want some kind of sparse matrix. Replacing the current dictionary with something like that would probably merit its own branch.

For the record, this branch doesn't change how the GLENUM lookup works; I just cleaned it up a bit and pulled its declaration out of the macro, and changed `GLENUM()` from having one method to having three.

Reply:
Sorry I forgot to answer. I did actually missunderstand your PR. I thought it was replacing the constants with `const GL_CONSTANT = GLENUM(...)`, which would put a lot more pressure on the implementation of `GLENUM`...

Reply:
Ah, that makes sense now :D

Well we've got the important bug-fixes merged in, and there's already an experimental branch using Clang, so it's up to you if you still want the other stuff from this PR. I'll try to fix the merge conflicts tonight (this branch should steamroll over the other one).

Either way, do you want some ARB extensions from another branch of mine? I only added a few that were relevant to me, mainly `ARB_bindless_textures`.